knowledge-base/adr/0024-url-encoding.md

```diff
@@ -0,0 +1,193 @@
+
+### URL
+
+```
+https://www.example.com/path?param=1
+```
+
+The example above consists of the following components:
+
+- protocol: `https` :thumbsdown:
+- host: `www.example.com` :thumbsdown:
+- path: `/path` :thumbsup:
+- query: `param=1` :thumbsup:
+
+As the protocol will not be encoded, we will only discuss the rest of them.
+
+### Host :thumbsdown:
+
+There are some cases where the host contains special characters like umlaut.
+As not all the browsers/http clients can handle such special characters, Internationalized Domain Name (IDN) plays a big role here.
+One can use an IDN converter so `www.häuser.de` can be converted to `www.xn--huser-gra.de`.
+As the host is usually configured in the destination, the SDK will not make any adjustment.
+
+### Path
+
+Let's discuss OData path and OpenAPI path separately, as OData case is a bit complicated.
+
+#### OData
+
+Below is one example about how the path should look like in terms of OData
+
+```
+/servicePath/Entity('key')/to_MultiLink('key')/to_SingleLink
+```
+
+- service path: `/servicePath` :thumbsdown:
+- entity path: `/Entity(KeyName='value')` :thumbsup:
+- navigation property path: `/to_MultiLink(KeyName='value')/to_SingleLink` :thumbsup:
+
+##### Service Path :thumbsdown:
+
+We assume the service path has not special characters that need to be encoded.
+If needed, the user can easily configure the service mapping file.
+
+##### Entity Path :thumbsup:
+
+The entity path consists of the following parts:
+
+- an entity name: `Entity` :thumbsdown:
+- entity key names: `KeyName` :thumbsdown:
+- values of the entity key: `'value'` :thumbsup:
+
+We assume the "entity names" and "entity key names" provided by the metadata do not have to be encoded.
+If needed, based on the feedback, we can either do it during generation or ask the possibility to adjust the metadata.
+
+On the other side, the "values of the entity key" of an entity is provided by the user, which might contain special characters.
+One of the users uses `/` as part of a key.
+Therefore, we have to encode the "values of the entity key".
+
+##### Navigation Property Path :thumbsup:
+
+Similar to entity path, only the "values of the navigation property key" are planned to be encoded.
+
+#### OpenAPI
+
+The path pattern should look like the example below:
+
+```
+"/entities/{entityId}"
+```
+
+- path: `/entities` :thumbsdown:
+- path parameter: `{entityId}` :thumbsup:
+
+Similar to OData, only user input is considered when encoding, which is the "path parameters"
+
+### Query
+
+Similar to path section, we also discuss OData and OpenAPI separately.
+
+#### OData
+
+OData allows you to use the following query options:
+
+- `$format=json` :thumbsdown:
+- `$top=1` :thumbsdown:
+- `$skip=1` :thumbsdown:
+- `$orderby=StringProperty` :thumbsdown:
+- `$select=StringProperty` :thumbsdown:
+- `$expand=to_MultiLink` :thumbsup:
+- `$filter=StringProperty eq 'string'` :thumbsup:
```

Comment:
Again, I find this misleading. I am not even sure whether we would encode $filter=**StringProperty eq 'string'** or $filter=StringProperty eq **'string'**, but this information here rather tells me that we would encode the whole thing.

Reply:
I think we encode `eq 'string'` so the empty spaces are encoded and the value put in.